README.rst
.. image:: sliceplots-two_dimensional-1.png
   :alt: Sample plot
   :width: 400px
   :target: sliceplots-two_dimensional-1.png


===========
Slice Plots
===========

.. image:: https://img.shields.io/pypi/v/sliceplots.svg
   :target: https://pypi.python.org/pypi/sliceplots

.. image:: https://img.shields.io/pypi/l/sliceplots.svg
   :target: https://github.com/berceanu/sliceplots/blob/master/LICENSE
   :alt: PyPI - License

Thin wrapper on top of ``matplotlib``'s ``.imshow()`` for 2D plotting, with attached slice plots

Features
--------

* only depends on ``matplotlib``
* uses its pure OO (look Ma, no ``pyplot``!) interface
* support for 1D and 2D plots with various customization options
* broken-axis 1D plots
* small codebase
* designed for non-interactive use, scripting and publication-quality plots


HISTORY.rst
=======
History
=======

0.3.3 (2023-10-11)
------------------

* Bring dependencies up to date

0.3.2 (2019-10-28)
------------------

* Implement vmin, vmax and alpha for colored line plot


0.3.1 (2019-08-06)
------------------

* Minor bugfixes and docs update.


0.3.0 (2019-08-03)
------------------

* Add colored line plot with colorbar.

0.2.1 (2019-08-02)
------------------

* Fix bumpversion problem

0.2.0 (2019-08-02)
------------------

* Simplify documentation and tests.
* Allow passing an existing ``Axes`` instance to all methods/classes.

0.1.2 (2019-06-19)
------------------

* Fixed label positions and colors.

0.1.0 (2019-06-17)
------------------

* Added regression tests, usage and docstrings.

0.0.1 (2019-06-16)
------------------

* First release on PyPI.


tests/conftest.py
import pytest
import numpy as np
from collections import namedtuple


@pytest.fixture(scope="module")
def plt_data():
    # data for plotting
    PlottingData = namedtuple("plotting_data", ["uu", "data", "idx"])
    uu = np.linspace(0, np.pi, 128)
    data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)
    return PlottingData(uu=uu, data=data, idx=data.shape[0] // 2)


tests/test_util.py
#!/usr/bin/env pytahon
# -*- coding: utf-8 -*-

"""Tests for `sliceplots._util` module."""

import numpy as np

from sliceplots.util import _idx_from_val


def test_idx_from_val():
    """Find element position in array."""
    position = _idx_from_val(np.linspace(0, 10, 11), 5.1)
    assert position == 5


tests/generate_figs.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import numpy as np

import sliceplots.two_dimensional as two_d

if __name__ == "__main__":
    axis_data = np.linspace(0, np.pi, 128)
    data_2d = np.cos(axis_data - 0.5) * np.cos(axis_data.reshape(-1, 1) - 1.0)

    p2d = two_d.Plot2D(
        data_2d,
        axis_data,
        axis_data,
        xlabel="x",
        ylabel="y",
        zlabel="f(x,y)",
        hslice_val=0.75,
        vslice_val=2.75,
        figsize=(6.0, 6.0),
    )
    p2d.fig.savefig("imshow_slices.png", bbox_inches="tight")


sliceplots/util.py
# -*- coding: utf-8 -*-

"""Utility functions module."""

import numpy as np

from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
from matplotlib.figure import Figure
from matplotlib import ticker
from mpl_toolkits.axes_grid1 import make_axes_locatable


def addcolorbar(*, ax, mappable, label=None, stub=False, **kwargs):
    r"""Add a colorbar to a matplotlib image.

    Parameters
    ----------
    ax : :class:`~matplotlib.axes.Axes`, list of Axes, optional
        Parent axes from which space for a new colorbar axes will be stolen.
        If a list of axes is given they will all be resized to make room for the
        colorbar axes.
    mappable : :class:`~matplotlib.cm.ScalarMappable`
        The :class:`~matplotlib.image.Image`,
        :class:`~matplotlib.contour.ContourSet`, etc.
        described by this colorbar.

    label : str, optional
        Colorbar axis label (defaults to ``None``).
    stub : bool, optional
        If ``True``, return the Axes into which the colorbar was drawn.
        The colorbar is invisible in this case. Defaults to ``False``.

    kwargs : dict, optional
        Keyword arguments that control the look of the colorbar.
        Optional keyword arguments include:

        ============= ====================================================
        Property      Description
        ============= ====================================================
        pos           position wrt parent axes: left, right, bottom or top
        size          5%; width, in percentage of the original axes width
        orientation   vertical or horizontal
        pad           0.05 if vertical, 0.15 if horizontal; fraction
                      of original axes between colorbar and new image axes
        max_ticks     maximum number of tick marks on the colorbar
        ============= ====================================================

    Returns
    -------
    cax : :class:`~matplotlib.axes.Axes`
        New axes instance, with drawn colorbar.

    Examples
    --------
    .. plot::
       :include-source:

        import numpy as np
        from matplotlib import pyplot

        from sliceplots import addcolorbar

        uu = np.linspace(0, np.pi, 128)
        data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)

        fig, ax = pyplot.subplots()
        img = ax.imshow(data)

        addcolorbar(ax=ax, mappable=img, label="x")
    """
    max_ticks = kwargs.pop("max_ticks", None)

    pos = kwargs.pop("pos", "right")
    size = kwargs.pop("size", "5%")
    orientation = kwargs.pop("orientation", "vertical")

    default_pad = 0.05
    if orientation == "horizontal":
        default_pad = 0.15
    pad = kwargs.pop("pad", default_pad)

    divider = make_axes_locatable(axes=ax)
    cax = divider.append_axes(position=pos, size=size, pad=pad)

    if stub:
        cax.set_visible(False)
        return cax

    cb = ax.figure.colorbar(mappable=mappable, cax=cax, orientation=orientation)

    if max_ticks is not None:
        tick_locator = ticker.MaxNLocator(nbins=max_ticks)
        cb.locator = tick_locator
        cb.update_ticks()

    cb.set_label(label)

    return cax


def _make_ax(**my_adjust):
    r"""

    Parameters
    ----------
    my_adjust : dict
        Parameters for tuning the subplot layout.

    Returns
    -------
    ax : :py:class:`matplotlib.axes.Axes`
        Axes instance, for plotting.

    Examples
    --------
    >>> _make_ax(left=0.15, right=0.98)  #doctest: +ELLIPSIS
    <Axes: >
    """
    adjust = dict(
        left=0.125,  # the left side of the subplots of the figure
        right=0.9,  # the right side of the subplots of the figure
        bottom=0.1,  # the bottom of the subplots of the figure
        top=0.9,  # the top of the subplots of the figure
        wspace=0.2,  # the amount of width reserved for space between subplots,
        # expressed as a fraction of the average axis width
        hspace=0.2,  # the amount of height reserved for space between subplots,
        # expressed as a fraction of the average axis height
    )

    if not my_adjust.keys() <= adjust.keys():
        raise TypeError("Unknown keyword arg!")

    adjust.update(my_adjust)

    fig = Figure()
    canvas = FigureCanvas(fig)
    canvas.figure.subplots_adjust(**adjust)
    ax = canvas.figure.add_subplot(111)

    return ax


def _idx_from_val(arr1d, val):
    """Given a 1D array, find index of closest element to given value.

    Parameters
    ----------
    arr1d : arr1d: :py:class:`numpy.ndarray`
        1D array of values.
    val : float
        Element value.

    Returns
    -------
    idx : int
        Element position in the array.

    Examples
    --------
    >>> import numpy as np
    >>> a = np.linspace(10, 20, 11)
    >>> _idx_from_val(a, 14.2)
    4
    """
    idx = (np.abs(arr1d - val)).argmin()
    return idx


sliceplots/one_dimensional.py
# -*- coding: utf-8 -*-

"""Module containing useful 1D plotting abstractions on top of matplotlib."""

import numpy as np
from mpl_toolkits.axes_grid1 import make_axes_locatable
from matplotlib.collections import LineCollection
from matplotlib.colors import Normalize

from sliceplots.util import _idx_from_val, _make_ax


def plot_multicolored_line(
    *,
    ax=None,
    x,
    y,
    other_y,
    cmap="viridis",
    vmin=None,
    vmax=None,
    linewidth=2,
    alpha=1
):
    r"""Plots a line colored based on the values of another array.

    Plots the curve ``y(x)``, colored based on the values in ``other_y``.

    Parameters
    ----------
    ax : :py:class:`~matplotlib.axes.Axes`, optional
        Axes instance, for plotting, defaults to ``None``.
        If ``None``, a new :py:class:`~matplotlib.figure.Figure` will be created.

    y : 1d array_like
        The dependent variable.
    x : 1d array_like
        The independent variable.
    other_y: 1d array_like
        The values whose magnitude will be converted to colors.

    cmap : str, optional
        The used colormap (defaults to "viridis").
    vmin : float, optional
        Lower normalization limit
    vmax : float, optional
        Upper normalization limit
    linewidth : float, optional (default 2)
        Width of the plotted line
    alpha : float, optional (default 1)
        Line transparency, between 0 and 1

    Returns
    -------
    ax, line : Axes, LineCollection
        Main Axes and plotted line

    Raises
    ------
    AssertionError
        If the length of `y` and `other_y` do not match.

    References
    ----------
    ``matplotlib`` `example <https://matplotlib.org/gallery/lines_bars_and_markers/multicolored_line.html>`_.

    Examples
    --------
    We plot a curve and color it based on the value of its first derivative.

    .. plot::
       :include-source:

        import numpy as np
        from matplotlib import pyplot

        from sliceplots import plot_multicolored_line

        x = np.linspace(0, 3 * np.pi, 500)
        y = np.sin(x)
        dydx = np.gradient(y) * 100  # first derivative

        _, ax = pyplot.subplots()

        plot_multicolored_line(ax=ax, x=x, y=y, other_y=dydx)

        ax.set(ylabel="y", xlabel="x")
    """
    if not (len(y) == len(other_y)):
        raise AssertionError("The two 'y' arrays must have the same size!")

    ax = ax or _make_ax()

    # Create a set of line segments so that we can color them individually
    # This creates the points as a N x 1 x 2 array so that we can stack points
    # together easily to get the segments. The segments array for line collection
    # needs to be (numlines) x (points per line) x 2 (for x and y)
    points = np.array([x, y]).T.reshape(-1, 1, 2)
    segments = np.concatenate([points[:-1], points[1:]], axis=1)

    if vmin is None:
        vmin = np.min(other_y)
    if vmax is None:
        vmax = np.max(other_y)

    # Create a continuous norm to map from data points to colors
    norm = Normalize(vmin, vmax)
    lc = LineCollection(segments, cmap=cmap, norm=norm)
    # Set the values used for colormapping
    lc.set_array(other_y)
    lc.set_linewidth(linewidth)
    lc.set_alpha(alpha)
    line = ax.add_collection(lc)

    return ax, line


def plot1d_break_x(*, ax=None, h_axis, v_axis, param, slice_opts):
    r"""Line plot with a broken x-axis.

    Parameters
    ----------
    ax : :py:class:`~matplotlib.axes.Axes`, optional
        Axes instance, for plotting.
        If ``None``, a new :py:class:`~matplotlib.figure.Figure` will be created.
        Defaults to ``None``.
    h_axis : 1d array_like
        x-axis data.
    v_axis : 1d array_like
        y-axis data.
    param : dict
        Axes limits and labels.
    slice_opts : dict
        Options for plotted line.

    Returns
    -------
    ax_left, ax_right : tuple of Axes
        Left and right Axes of the split plot.

    Examples
    --------
    .. plot::
        :include-source:

            import numpy as np
            from matplotlib import pyplot

            from sliceplots import plot1d_break_x

            uu = np.linspace(0, np.pi, 128)
            data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)

            _, ax = pyplot.subplots()

            plot1d_break_x(
                ax=ax,
                h_axis=uu,
                v_axis=data[data.shape[0] // 2, :],
                param={
                    "xlim_left": (0, 1),
                    "xlim_right": (2, 3),
                    "xlabel": r"$x$ ($\mu$m)",
                    "ylabel": r"$\rho$ (cm${}^{-3}$)",
                },
                slice_opts={"ls": "--", "color": "#d62728"})

    """
    ax_left = ax or _make_ax()

    divider = make_axes_locatable(ax_left)
    ax_right = divider.new_horizontal(size="100%", pad=1)
    ax_left.figure.add_axes(ax_right)

    ax_left.plot(h_axis, v_axis, **slice_opts)
    ax_left.set_ylabel(param["ylabel"])
    ax_left.set_xlabel(param["xlabel"])

    ax_left.set_xlim(*param["xlim_left"])
    ax_left.spines["right"].set_visible(False)
    ax_left.yaxis.set_ticks_position("left")

    ax_right.plot(h_axis, v_axis, **slice_opts)
    ax_right.set_ylabel(param["ylabel"])
    ax_right.set_xlabel(param["xlabel"])
    ax_right.yaxis.set_label_position("right")

    ax_right.set_xlim(*param["xlim_right"])
    ax_right.spines["left"].set_visible(False)
    ax_right.yaxis.set_ticks_position("right")

    # From https://matplotlib.org/examples/pylab_examples/broken_axis.html
    d = 0.015  # how big to make the diagonal lines in axes coordinates
    # arguments to pass plot, just so we don't keep repeating them
    kwargs = dict(transform=ax_left.transAxes, color="k", clip_on=False)
    ax_left.plot((1 - d, 1 + d), (-d, +d), **kwargs)
    ax_left.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)

    kwargs.update(transform=ax_right.transAxes)  # switch to the right axes
    ax_right.plot((-d, +d), (1 - d, 1 + d), **kwargs)
    ax_right.plot((-d, +d), (-d, +d), **kwargs)

    return ax_left, ax_right


def plot1d(*, ax=None, h_axis, v_axis, xlabel=r"", ylabel=r"", **kwargs):
    r"""Plot the data with given labels and plot options.

    Parameters
    ----------
    ax : class:`~matplotlib.axes.Axes`, optional
        Axes instance, for plotting.
        If ``None``, a new :class:`~matplotlib.figure.Figure` will be created.
        Defaults to ``None``.

    h_axis : :py:class:`np.ndarray`
        x-axis data.
    v_axis : :py:class:`np.ndarray`
        y-axis data.
    xlabel : str, optional
        x-axis label.
    ylabel : str, optional
        y-axis label.

    kwargs : dict, optional
        Other arguments for :meth:`~matplotlib.axes.Axes.plot`.

    Returns
    -------
    ax : Axes
        Modified Axes, containing plot.

    Examples
    --------
    >>> import numpy as np

    >>> uu = np.linspace(0, np.pi, 128)
    >>> data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)

    >>> plot1d(
    ...     h_axis=uu,
    ...     v_axis=data[data.shape[0] // 2, :],
    ...     xlabel=r"$z$ ($\mu$m)",
    ...     ylabel=r"$a_0$",
    ...     xlim=[0, 3],
    ...     ylim=[-1, 1],
    ...     color="#d62728",
    ... )  #doctest: +ELLIPSIS
    <Axes: xlabel='$z$ ($\\mu$m)', ylabel='$a_0$'>
    """
    xlim = kwargs.pop("xlim", [np.min(h_axis), np.max(h_axis)])
    ylim = kwargs.pop("ylim", [np.min(v_axis), np.max(v_axis)])
    #
    xmin_idx, xmax_idx = (
        _idx_from_val(h_axis, xlim[0]),
        _idx_from_val(h_axis, xlim[1]),
    )
    #
    h_axis = h_axis[xmin_idx:xmax_idx]
    data = v_axis[xmin_idx:xmax_idx]
    #
    label = {"x": xlabel, "y": ylabel}
    text = kwargs.pop("text", "")
    #
    ax = ax or _make_ax()

    ax.plot(h_axis, data, **kwargs)

    ax.set(
        xlim=[h_axis[0], h_axis[-1]], ylim=ylim, ylabel=label["y"], xlabel=label["x"]
    )

    ax.text(0.02, 0.95, text, transform=ax.transAxes, color="firebrick")

    return ax


sliceplots/sliceplots.py
# -*- coding: utf-8 -*-

"""Main module, containing useful plotting abstractions on top of matplotlib."""

import numpy as np
from matplotlib.artist import setp, getp
from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas
from matplotlib.figure import Figure
from matplotlib.gridspec import GridSpec
from mpl_toolkits.axes_grid1 import make_axes_locatable
from mpl_toolkits.axes_grid1.inset_locator import inset_axes


def idx_from_val(arr1d, val):
    """Given a 1D array `arr1d`, find index of closest value to `val`."""
    idx = (np.abs(arr1d - val)).argmin()
    return idx


class Plot2D:
    """
    Pseudocolor plot of a 2D array with optional 1D slices attached.
    """

    def __init__(
        self,
        arr2d: np.ndarray,
        h_axis: np.ndarray,
        v_axis: np.ndarray,
        xlabel=r"",
        ylabel=r"",
        zlabel=r"",
        **kwargs
    ) -> None:
        r"""
        >>> uu = np.linspace(0, np.pi, 128)
        >>> data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)
        >>> p2d = Plot2D(data, uu, uu,
                    xlabel=r'$x$ ($\mu$m)', ylabel=r'$y$ ($\mu$m)', zlabel=r'$\rho$ (cm$^{-3}$)',
                    hslice_val=0,
                    vslice_val=840.5,
                    hslice_opts={'color': 'firebrick', 'lw' : 0.5, 'ls':'-'},
                    vslice_opts={'color': 'blue', 'ls': '-'},
                    figsize=(8, 8), cmap='viridis', cbar=True,
                    extent=(822, 865, -20, 20),
                    vmin=-0.1, vmax=10,
                    text='iteration = {}'.format(35100),
                    norm=colors.SymLogNorm(linthresh=1e-4)
                    )
        Other options for norm:
        >>> import matplotlib.colors as colors
        >>> norm = colors.LogNorm()

        :param arr2d: data to be plotted
        :param h_axis: values on the "x" axis
        :param v_axis: values on the "y" axis
        :param xlabel: "x" axis label
        :param ylabel: "y" axis label
        :param zlabel: colorbar label
        :param kwargs: other arguments for ``matplotlib``
        """
        self.extent = kwargs.get(
            "extent", (np.min(h_axis), np.max(h_axis), np.min(v_axis), np.max(v_axis))
        )
        #
        xmin, xmax, ymin, ymax = self.extent
        xmin_idx, xmax_idx = idx_from_val(h_axis, xmin), idx_from_val(h_axis, xmax)
        ymin_idx, ymax_idx = idx_from_val(v_axis, ymin), idx_from_val(v_axis, ymax)
        #
        self.data = arr2d[ymin_idx:ymax_idx, xmin_idx:xmax_idx]
        self.min_data, self.max_data = np.amin(self.data), np.amax(self.data)
        self.vmin, self.vmax = (
            kwargs.get("vmin", self.min_data),
            kwargs.get("vmax", self.max_data),
        )
        #
        self.h_axis = h_axis[xmin_idx:xmax_idx]
        self.v_axis = v_axis[ymin_idx:ymax_idx]
        # np.clip(self.data, self.vmin, self.vmax, self.data)
        #
        self.label = {"x": xlabel, "y": ylabel, "z": zlabel}
        #
        self.cbar = kwargs.get("cbar", True)
        # see https://matplotlib.org/users/colormapnorms.html
        self.norm = kwargs.get("norm")
        #
        self.hslice_val = kwargs.get("hslice_val")
        self.vslice_val = kwargs.get("vslice_val")
        self.hslice_idx = None
        self.vslice_idx = None
        if self.hslice_val is not None:
            self.hslice_idx = idx_from_val(self.v_axis, self.hslice_val)
        if self.vslice_val is not None:
            self.vslice_idx = idx_from_val(self.h_axis, self.vslice_val)
            #
        #
        self.text = kwargs.get("text", "")
        #
        self.fig = Figure(figsize=kwargs.pop("figsize", (8, 8)))
        # A canvas must be manually attached to the figure (pyplot would automatically
        # do it).  This is done by instantiating the canvas with the figure as
        # argument.
        self.canvas = FigureCanvas(self.fig)
        #
        self.draw_fig(**kwargs)

    def __str__(self):
        return "extent=({:.3f}, {:.3f}, {:.3f}, {:.3f}); min, max = ({:.3f}, {:.3f})".format(
            np.min(self.h_axis),
            np.max(self.h_axis),
            np.min(self.v_axis),
            np.max(self.v_axis),
            np.amin(self.data),
            np.amax(self.data),
        )

    @staticmethod
    def colorbar(mappable):
        r"""Constructs a scaled colorbar for a given plot.

        Parameters
        ----------
        mappable : The Image, ContourSet, etc. to which the colorbar applies.
        """
        ax = mappable.axes
        fig = ax.figure
        divider = make_axes_locatable(ax)
        cax = divider.append_axes("right", size="5%", pad=0.05)
        return fig.colorbar(mappable, cax=cax)

    def main_panel(self, **kwargs):
        self.im = self.ax0.imshow(
            self.data,
            origin="lower",
            extent=self.extent,
            aspect="auto",
            norm=self.norm,
            interpolation="none",
            cmap=kwargs.get("cmap", "viridis"),
            vmin=self.vmin,
            vmax=self.vmax,
        )
        #
        self.ax0.set_xlabel(self.label["x"])
        self.ax0.set_ylabel(self.label["y"])

    def draw_fig(self, **kwargs):
        slice_opts = {"ls": "-", "color": "firebrick", "lw": 0.5}  # defaults
        hslice_opts = slice_opts.copy()
        vslice_opts = slice_opts.copy()
        #
        hslice_opts.update(kwargs.get("hslice_opts", {}))
        vslice_opts.update(kwargs.get("vslice_opts", {}))

        # #
        if (self.hslice_idx is None) and (self.vslice_idx is None):
            gs = GridSpec(1, 1, height_ratios=[1], width_ratios=[1])
            self.ax0 = self.fig.add_subplot(gs[0])
            self.main_panel(**kwargs)

        # ---- #
        elif (self.hslice_idx is not None) and (self.vslice_idx is None):
            gs = GridSpec(2, 1, height_ratios=[1, 3], width_ratios=[1])
            self.ax0 = self.fig.add_subplot(gs[1, 0])
            self.axh = self.fig.add_subplot(gs[0, 0], sharex=self.ax0)
            #
            self.main_panel(**kwargs)
            #
            self.ax0.axhline(y=self.v_axis[self.hslice_idx], **hslice_opts)
            #
            self.ax0.annotate(
                "{:.1f}".format(self.v_axis[self.hslice_idx]),
                xy=(self.h_axis[3], self.v_axis[self.hslice_idx + 3]),
                xycoords="data",
                color=hslice_opts["color"],
            )
            #
            self.axh.set_xmargin(0)
            self.axh.set_ylabel(self.label["z"])
            self.axh.plot(self.h_axis, self.data[self.hslice_idx, :], **hslice_opts)
            self.axh.set_ylim(self.vmin, self.vmax)
            #
            self.axh.xaxis.set_visible(False)
            #
            for sp in ("top", "bottom", "right"):
                self.axh.spines[sp].set_visible(False)
            #
            self.fig.subplots_adjust(hspace=0.03)

        # | #
        elif (self.vslice_idx is not None) and (self.hslice_idx is None):
            gs = GridSpec(1, 2, height_ratios=[1], width_ratios=[3, 1])
            self.ax0 = self.fig.add_subplot(gs[0, 0])
            self.axv = self.fig.add_subplot(gs[0, 1], sharey=self.ax0)
            #
            self.main_panel(**kwargs)
            #
            self.ax0.axvline(x=self.h_axis[self.vslice_idx], **vslice_opts)
            #
            self.ax0.annotate(
                "{:.1f}".format(self.h_axis[self.vslice_idx]),
                xy=(self.h_axis[self.vslice_idx - 40], self.v_axis[-40]),
                xycoords="data",
                color=vslice_opts["color"],
                rotation="vertical",
            )
            #
            self.axv.set_ymargin(0)
            self.axv.set_xlabel(self.label["z"])
            self.axv.plot(self.data[:, self.vslice_idx], self.v_axis, **vslice_opts)
            self.axv.set_xlim(self.vmin, self.vmax)
            #
            self.axv.yaxis.set_visible(False)
            #
            for sp in ("top", "left", "right"):
                self.axv.spines[sp].set_visible(False)
            #
            self.fig.subplots_adjust(wspace=0.03)

        # --|-- #
        else:
            gs = GridSpec(2, 2, height_ratios=[1, 3], width_ratios=[3, 1])
            self.ax0 = self.fig.add_subplot(gs[1, 0])
            self.axh = self.fig.add_subplot(gs[0, 0], sharex=self.ax0)
            self.axv = self.fig.add_subplot(gs[1, 1], sharey=self.ax0)
            #
            self.main_panel(**kwargs)
            #
            self.ax0.axhline(y=self.v_axis[self.hslice_idx], **hslice_opts)  # ##----##
            self.ax0.axvline(x=self.h_axis[self.vslice_idx], **vslice_opts)  # ## | ##
            # --- #
            self.ax0.annotate(
                "{:.1f}".format(self.v_axis[self.hslice_idx]),
                xy=(self.h_axis[3], self.v_axis[self.hslice_idx + 3]),
                xycoords="data",
                color=hslice_opts["color"],
            )
            # | #
            self.ax0.annotate(
                "{:.1f}".format(self.h_axis[self.vslice_idx]),
                xy=(self.h_axis[self.vslice_idx - 40], self.v_axis[-40]),
                xycoords="data",
                color=vslice_opts["color"],
                rotation="vertical",
            )
            # --- #
            self.axh.set_xmargin(0)  # otherwise ax0 may have white margins
            self.axh.set_ylabel(self.label["z"])
            self.axh.plot(self.h_axis, self.data[self.hslice_idx, :], **hslice_opts)
            self.axh.set_ylim(self.vmin, self.vmax)
            # self.axh.set_yticks([-1, 0, 1])
            # | #
            self.axv.set_ymargin(0)
            self.axv.set_xlabel(self.label["z"])
            self.axv.plot(self.data[:, self.vslice_idx], self.v_axis, **vslice_opts)
            self.axv.set_xlim(self.vmin, self.vmax)
            # hide the relevant axis
            self.axh.xaxis.set_visible(False)  # -
            self.axv.yaxis.set_visible(False)  # |
            # "Despine" the slice profiles
            for ax, spines in (
                (self.axh, ("top", "bottom", "right")),
                (self.axv, ("top", "left", "right")),
            ):
                #
                for sp in spines:
                    ax.spines[sp].set_visible(False)
            #
            self.fig.subplots_adjust(wspace=0.03, hspace=0.03)

        # self.fig.tight_layout()
        #
        self.ax0.text(
            0.02, 0.02, self.text, transform=self.ax0.transAxes, color="firebrick"
        )
        #
        if self.cbar:
            cax = inset_axes(self.ax0, width="70%", height="3%", loc=2)
            cbar = self.fig.colorbar(
                self.im, cax=cax, orientation="horizontal"
            )  # ticks=[self.vmin, self.vmax]
            # cbar.set_label(self.label['z'], color='firebrick')
            self.ax0.text(
                0.74,
                0.97,
                self.label["z"],
                transform=self.ax0.transAxes,
                color="firebrick",
            )
            # cbar.ax.xaxis.set_ticks_position('top')
            # cbar.ax.xaxis.set_label_position('top')
            cbar.ax.tick_params(color="firebrick", width=1.5, labelsize=8)
            cbxtick_obj = getp(cbar.ax.axes, "xticklabels")
            setp(cbxtick_obj, color="firebrick")


class Plot1D:
    """
    Plot of 1D array.
    """

    def __init__(
        self, arr1d: np.ndarray, h_axis: np.ndarray, xlabel=r"", ylabel=r"", **kwargs
    ) -> None:
        r"""
        >>> plot = Plot1D(a0, z0, xlabel=r'$%s \;(\mu m)$'%'z', ylabel=r'$%s$'%'a_0',
                                xlim=[0, 900], ylim=[0, 10],
                                figsize=(10, 6), color='red')
        >>> plot.canvas.print_figure('a0.png')

        :param arr1d: data to be plotted on the "y" axis
        :param h_axis: values on the "x" axis
        :param xlabel: "x" axis label
        :param ylabel: "y" axis label
        :param kwargs: other arguments for ``matplotlib.plot()``
        """
        self.xlim = kwargs.pop("xlim", [np.min(h_axis), np.max(h_axis)])
        self.ylim = kwargs.pop("ylim", [np.min(arr1d), np.max(arr1d)])
        #
        xmin_idx, xmax_idx = (
            idx_from_val(h_axis, self.xlim[0]),
            idx_from_val(h_axis, self.xlim[1]),
        )
        #
        self.h_axis = h_axis[xmin_idx:xmax_idx]
        self.data = arr1d[xmin_idx:xmax_idx]
        #
        self.label = {"x": xlabel, "y": ylabel}
        self.text = kwargs.pop("text", "")
        #
        self.fig = Figure(figsize=kwargs.pop("figsize", (6.4, 6.4)))
        self.canvas = FigureCanvas(self.fig)
        self.ax = self.fig.add_subplot(111)

        self.ax.plot(self.h_axis, self.data, **kwargs)

        self.ax.set(
            xlim=[self.h_axis[0], self.h_axis[-1]],
            ylim=self.ylim,
            ylabel=self.label["y"],
            xlabel=self.label["x"],
        )

        self.ax.grid()

        self.ax.text(
            0.02, 0.95, self.text, transform=self.ax.transAxes, color="firebrick"
        )

    def __str__(self):
        return "extent=({:.3f}, {:.3f}); min, max = ({:.3f}, {:.3f})".format(
            np.min(self.h_axis),
            np.max(self.h_axis),
            np.amin(self.data),
            np.amax(self.data),
        )


def plot1d_break_x(fig, h_axis, v_axis, param, slice_opts):
    r"""
    >>> uu = np.linspace(0, np.pi, 128)
    >>> data = np.cos(uu - 0.5) * np.cos(uu.reshape(-1, 1) - 1.0)
    >>> fig, ax = plt.subplots(figsize=(8, 3.2))
    >>> plot1d_break_x(fig, uu, data[data.shape[0]//2], {'xlim_left':(0,1), 'xlim_right':(2,3),
        'xlabel':r'$x$ ($\mu$m)', 'ylabel':r'$\rho$ (cm$^{-3}$)'}, {'ls': '--', 'color': '0.5'})
    >>> fig
    """
    ax_left = fig.axes[0]
    divider = make_axes_locatable(ax_left)
    ax_right = divider.new_horizontal(size="100%", pad=1)
    fig.add_axes(ax_right)

    ax_left.plot(h_axis, v_axis, **slice_opts)
    ax_left.set_ylabel(param["ylabel"])
    ax_left.set_xlabel(param["xlabel"])

    ax_left.set_xlim(*param["xlim_left"])
    ax_left.yaxis.tick_left()
    ax_left.tick_params(labelright="off")
    ax_left.spines["right"].set_visible(False)

    ax_right.plot(h_axis, v_axis, **slice_opts)
    ax_right.set_ylabel(param["ylabel"])
    ax_right.set_xlabel(param["xlabel"])
    ax_right.yaxis.set_label_position("right")

    ax_right.set_xlim(*param["xlim_right"])
    ax_right.yaxis.tick_right()
    ax_right.spines["left"].set_visible(False)

    # From https://matplotlib.org/examples/pylab_examples/broken_axis.html
    d = 0.015  # how big to make the diagonal lines in axes coordinates
    # arguments to pass plot, just so we don't keep repeating them
    kwargs = dict(transform=ax_left.transAxes, color="k", clip_on=False)
    ax_left.plot((1 - d, 1 + d), (-d, +d), **kwargs)
    ax_left.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)

    kwargs.update(transform=ax_right.transAxes)  # switch to the right axes
    ax_right.plot((-d, +d), (1 - d, 1 + d), **kwargs)
    ax_right.plot((-d, +d), (-d, +d), **kwargs)
